13 use validators for the models

SantiagoTorres opened the pull request:
This patch series refactors code in the models and adds the validate function to ensure the properties of a model are populated properly. I also added unit tests for these methods.

closes #13

SantiagoTorres committed:
ENH: toto/formats: remove, as it is no longer used

SantiagoTorres committed:
FIX: models: layout: the keys property should be a dict

Fix the devault value for the keys property. It should be a dict

SantiagoTorres committed:
ENH: toto/models/validators: refactor module

The model validator module was pretty bloated and useless overall.
Remove the unnecessary logic and refactor functions to improve
readability.

SantiagoTorres committed:
ENH:toto:models:validators: rename to matchrule module

The validators module only contained functions relevant to the
matchrules. Rename the module to "matchrule," which is a better fitting
name.

SantiagoTorres committed:
ENH: models:matchrule: add unit tests, update behavior

The matchrule syntax checker module didn't account for the "AS
<new_path>" construction. Add the relevant code to handle this. In
addition, add unit tests to ensure the module's correctness.

SantiagoTorres committed:
ENH: models: add validator function

The current models do not ensure that their properties are properly
populated. This may make creating and dumping layouts error-prone.

Add a validate method for the metadata objects that ensures that each
field/property is correctly populated.

lukpueh commented:
cdc03038575eb7742dcb21ec5fea62fc955ccf2b should not be part of this PR

lukpueh commented:
I think we agreed to [avoid trailing whitespaces](https://github.com/secure-systems-lab/code-style-guidelines/issues/5). Mind stripping those?

lukpueh commented on toto/models/common.py:
```diff
@@ -43,6 +44,14 @@ def dump(self, filename):
       fp.write("{}".format(self))
 
+  def validate(self):
+    """ Aggregates all the validate methods of itself and its subclasses """
```
This could use a little more information, à la *purpose, arguments, exceptions, side effects, and return value*. Especially since it those *esoteric* stuff like class inspection. :)

SantiagoTorres committed:
STY: fix trailing whitespace

There were some trailing whitespaces from the previous commits. I
removed them to match the coding guidelines

SantiagoTorres commented:
Fixed

lukpueh commented on toto/models/layout.py:
```diff
@@ -136,6 +143,62 @@ def import_step_metadata_from_files_as_dict(self):
     return step_link_dict
 
+  def _validate_type(self):
+    """ ensure that the type string is set to layout"""
```
I think your oneliner docstrings should look like this, according to our code style guidelines:
```python
"""Ensure that the type string is set to layout. """
```
Please change everywhere in this PR.

lukpueh commented on toto/models/layout.py:
```diff
@@ -136,6 +143,62 @@ def import_step_metadata_from_files_as_dict(self):
+  def _validate_expires(self):
+    """ ensure that the expiration field is a proper timestamp"""
+    try:
+      date = parse(self.expires)
```
This should be more restrictive. The required schema is [ISO8601_DATETIME_SCHEMA](https://github.com/secure-systems-lab/ssl_crypto/blob/master/formats.py#L94).

lukpueh commented on toto/models/layout.py:
```diff
@@ -136,6 +143,62 @@ def import_step_metadata_from_files_as_dict(self):
+
+    for keyid in self.keys:
+      toto.ssl_crypto.formats.KEYID_SCHEMA.check_match(keyid)
+      toto.ssl_crypto.formats.ANYKEY_SCHEMA.check_match(self.keys[keyid])
```
Do we need a comment here or in the docstrings that this raises `FormatError` in the case of not matching?

lukpueh replied:
Btw. there is a [KEYDICT_SCHEMA](https://github.com/secure-systems-lab/ssl_crypto/blob/master/formats.py#L352) already combining these two.

lukpueh commented on toto/models/layout.py:
```diff
@@ -136,6 +143,62 @@ def import_step_metadata_from_files_as_dict(self):
+    if type(self.keys) != dict:
+      raise FormatError("keys dictionary is malformed!")
+
+    for keyid in self.keys:
```
Let's use `iteritems` to iterate over dictionaries. And let's use it with `six`, to prepare for python2/3 compatibility.

lukpueh commented on toto/models/common.py:
```diff
@@ -43,6 +44,14 @@ def dump(self, filename):
+          if not method[1]():
+            return False
+
+    return True
```
I don't think we should mix `True`/`False` and `Exception`/`None` styles here. As far as I understand the methods return either `True` if validation passes or raise an `Exception` else. I think we can omit the bool retruns here and in the `_validate_*` methods.

lukpueh commented on toto/models/layout.py:
```diff
@@ -136,6 +143,62 @@ def import_step_metadata_from_files_as_dict(self):
+      inspection.validate()
+
+    return True
+
```
Maybe it's worth adding a *validate step and inspection name uniqueness* part here?

lukpueh commented on toto/models/layout.py:
```diff
@@ -176,6 +239,49 @@ def read(data):
+
+  def _validate_expected_command(self):
+    """ check that the expected command is a command-like string"""
+    # TODO: Maybe be a little bit more granularity
```
What do you mean by granularity? I think this is fine, even if we decide to change this to [`expected_command_pattern`](https://github.com/in-toto/in-toto/issues/46), checking for a string should be fine.

lukpueh commented on toto/models/layout.py:
```diff
@@ -207,3 +313,30 @@ def read(data):
+
+  def _validate_run(self):
+    """ check that the expected command is a command-like string"""
+    # TODO: Maybe be a little bit more granularity
```
See comment above. I think this is okay!

lukpueh commented on toto/models/matchrule.py:
```diff
@@ -0,0 +1,102 @@
+"""
+<Module Name>
+  toto/models/validators.py
```
Is this `toto/models/validators.py` or `toto/models/matchrule.py`? I think `matchrule_validators.py` would make sense.

lukpueh commented on toto/models/matchrule.py:
```diff
@@ -0,0 +1,102 @@
+  MATERIAL_OR_PRODUCT = {'PRODUCT', 'MATERIAL'}
+
+  if not isinstance(keywords, list):
+    raise FormatError("this matching rule is not a list")
```
Btw. I think we should allow the rule to be written as strings. A `.split()` when parsing the rule is not very expensive, whereas a split when writing the rule somehow feels stupid. I for instance script my rules like so: 
```python
material_matchrules = ["MATCH PRODUCT yadda yadda".split(), ...]
```

SantiagoTorres replied:
The way you do it is fine. But for the sake of the validator (that checks what should be in the metadata) , I think we should explicitly require lists.

lukpueh replied:
👍  Let's delegate this issue: https://github.com/in-toto/in-toto/issues/51

lukpueh commented on toto/models/matchrule.py:
```diff
@@ -0,0 +1,102 @@
+
+  <Raises>
+    FormatError: if the keywords provided do not match the matchrule syntax
+  """
```
Please stick to the code style guidelines: *purpose, arguments, exceptions, side effects, and return value*.

lukpueh replied:
Ping

lukpueh commented on toto/models/matchrule.py:
```diff
@@ -0,0 +1,102 @@
+    raise FormatError("this matching rule is not a list")
+
+  if len(keywords) == 5:
+    rule, artifact, target, from_keyword, step = keywords
```
I've only seen this with list returning functions. But it looks nifty with given lists too. :)

lukpueh commented on toto/models/matchrule.py:
```diff
@@ -0,0 +1,102 @@
+
+def _validate_match_rule(keywords):
+  """ private helper to verify the syntax of the MATCH matchrule """
+  MATERIAL_OR_PRODUCT = {'PRODUCT', 'MATERIAL'}
```
Is there a reason why you initialize this as a `set`? If so, I'd prefer to not use the literal set notation. I think it's easily confused with a dictionary.

SantiagoTorres replied:
I initialize this as a set because it is faster to check for membership and easier to extend. Which notation would you prefer?

lukpueh replied:
Got it. Actually, I'm fine with this notation.

lukpueh commented on toto/models/matchrule.py:
```diff
@@ -0,0 +1,102 @@
+def _validate_generic_rule(keywords):
+  """ private helper that verifies the syntax of the other rules """
+
+  VALID_OPERATIONS = {'CREATE', 'MODIFY', 'DELETE',}
```
See comment [above](https://github.com/in-toto/in-toto/pull/48/files#diff-7f9b331782acd3d05b10f3ab0a7609f0R24).

lukpueh commented on toto/models/matchrule.py:
```diff
@@ -0,0 +1,102 @@
+
+  rule, artifact = keywords
+
+  toto.ssl_crypto.formats.PATH_SCHEMA.check_match(artifact)
```
Actually this can be a *path pattern*. Since [`PATH_SCHEMA`](https://github.com/secure-systems-lab/ssl_crypto/blob/master/formats.py#L123) only requires it to be any string, we are fine here.

lukpueh commented on toto/models/matchrule.py:
```diff
@@ -0,0 +1,102 @@
+    new_target = target
+
+  elif len(keywords) == 7:
+    rule, artifact, target, from_keyword, step, as_keyword, new_target = keywords
```
Please take a look at the current [matchrule implementation](https://github.com/in-toto/in-toto/blob/develop/toto/verifylib.py#L280-L285)
It's actually:
```
MATCH (MATERIAL|PRODUCT) <path pattern>  [AS <target path pattern>] FROM <step name>
```

lukpueh replied:
This also affects the rest of this function and the unittests.

SantiagoTorres committed:
FIX: test:models: missing __init__.py file

Without this file, the test discovery script won't find the models
subdirectory.

lukpueh commented on toto/models/layout.py:
```diff
@@ -207,3 +313,30 @@ def read(data):
+  def _validate_type(self):
+    """ ensure that the type field is set to inspection"""
+    if self._type != "inspection":
+      raise FormatError("The _type field should be aset to inspection!")
```
Tiny typo.

lukpueh commented on test/models/test_inspection.py:
```diff
@@ -0,0 +1,93 @@
+    with self.assertRaises(FormatError):
+      self.inspection._type = "wrong"
+      self.inspection._validate_type()
+      self.assertFalse(self.inspection.validate())
```
I don't understand this. You never reach the `assertFalse` because the `FormatError` gets raised before. Please see my comment [below](https://github.com/in-toto/in-toto/pull/48/files#diff-cc71a9e93777c3c30e19a468fd9fede4R54) regarding Exception vs Bool strategy. The comment applies for most of the unittests below.

lukpueh commented on test/models/test_inspection.py:
```diff
@@ -0,0 +1,93 @@
+    self.assertFalse(self.inspection.validate())
+
+  def test_wrong_run(self):
+
```
A onliner docstring would help here.

lukpueh commented on test/models/test_layout.py:
```diff
@@ -0,0 +1,134 @@
+  def setUp(self):
+    """ Populate a base layout that we can use """
+    self.layout = Layout()
+    self.layout.expires = datetime.datetime.now().isoformat()
```
Please check out this [comment]( https://github.com/in-toto/in-toto/pull/48/files#diff-03000fa12a2e142e6c48c5bff130c268R156).

lukpueh commented on test/models/test_inspection.py:
```diff
@@ -0,0 +1,93 @@
+
+      self.inspection.material_matchrules = "PFF"
+      self.inspection._validate_material_matchrules()
+      self.assertFalse(self.inspection.validate())
```
You only get until [line 48](https://github.com/in-toto/in-toto/pull/48/files#diff-ba23fca1006b681698d19e43ee88da7eR48). You need a separate `with self.assertRaises(FormatError)` for each exception that gets raised. This is a re-occurring thing in the unit tests below.

lukpueh commented on test/models/test_layout.py:
```diff
@@ -0,0 +1,134 @@
+    with self.assertRaises(FormatError):
+
+      # FIXME: attr.ib reutilizes the default dictionary, so future constructor
+      # are not empty...
```
We should think about getting rid of `attr.ib`.

lukpueh commented on test/models/test_matchrule.py:
```diff
@@ -0,0 +1,77 @@
+    check_matchrule_syntax(rule)
+
+    rule = ["MATCH", "PRODUCT", "foo", "FROM", "step", "AS", "bar"]
+    check_matchrule_syntax(rule)
```
Please see [this comment](https://github.com/in-toto/in-toto/pull/48/files#diff-7f9b331782acd3d05b10f3ab0a7609f0R35).

lukpueh reviewed (changes requested):
Thanks for all the changes. Overall, very smart and clean code. Please consider my comments and we are good to go.

lukpueh commented:
Btw. I wished that this PR would also add validation calls throughout the codebase where appropriate. Eg. lines like [`# FIXME: Validate rule format`](https://github.com/in-toto/in-toto/blob/develop/toto/verifylib.py#L324)

lukpueh commented:
Overall, clean and smart code and  much appreciated additions, @SantiagoTorres. Thanks for the PR!

SantiagoTorres commented:
@lukpueh let me add a commit with your comments and another addressing the FIXME's

SantiagoTorres committed:
FIX: toto/models: code style fixes and minor patches

Addresses comments by Lukas

- Fixes docstrings throughout the test modules

- Adds more strict checks for the date time field

- Uses proper schema matchers to check key dictionaries

- Fixes the matchrule syntax and relocates the module to a more sensible
  place

- Fixes wrong context processors when assertRaises it uses

- Removes outdated comments

SantiagoTorres committed:
ENH: verifylib: uses the matchrule_validators.

Throughout verifylib the check_matchrule_syntax is added to ensure the
syntax is proper.

SantiagoTorres commented:
@lukpueh ready for another go

SantiagoTorres committed:
STY: clears whitespace errors

SantiagoTorres committed:
ENH: models/layout.Layout: check repeated names

It may be possible that the Layout instances have two or more instances
of Step or Inspection objects with the same name. This produces an
ambiguous and invalid layout. We add a check to avoid duplicate names
within the validate methods for the Layout instance.

lukpueh commented on toto/models/layout.py:
```diff
@@ -197,7 +196,10 @@ def _validate_inspections(self):
 
-    return True
+      if inspection.name in names_seen:
+        raise FormatError("There is a repeated name in the steps! "
```
```python
FormatError("There is a repeated name in the inspections")
```

lukpueh commented on toto/models/layout.py:
```diff
@@ -315,28 +317,27 @@ def read(data):
-    """ ensure that the type field is set to inspection"""
+    """Private method to ensure that the type field is set to inspection."""
     if self._type != "inspection":
       raise FormatError("The _type field should be aset to inspection!")
```
Please change "aset" to "set"

lukpueh commented on toto/matchrule_validators.py:
```diff
@@ -32,22 +32,24 @@ def _validate_match_rule(keywords):
 
   elif len(keywords) == 7:
-    rule, artifact, target, from_keyword, step, as_keyword, new_target = keywords
+    (rule, artifact, target, as_keyword, new_target, from_keyword,
```
You call the left path argument "target" and the right path argument "new_target" which is different from the existing naming scheme in verifylib.
```python
["MATCH", "MATERIAL", "<path pattern>", "FROM", "<step name>"]
["MATCH", "PRODUCT", "<path pattern>", "FROM", "<step name>"]
["MATCH", "MATERIAL", "<path pattern>", "AS", "<target path pattern>", "FROM", "<step name>"]
["MATCH", "PRODUCT", "<path pattern>", "AS", "<target path pattern>", "FROM", "<step name>"]
```
I'm not happy with either of them but I think we should stay consistent, or at least add a `# FIXME` comment.

lukpueh commented on toto/matchrule_validators.py:
```diff
@@ -65,7 +67,7 @@ def _validate_generic_rule(keywords):
+  if rule not in VALID_OPERATIONS and rule.upper() not in VALID_OPERATIONS:
     raise FormatError("{} is not a valid rule!".format(rule))
 
 def check_matchrule_syntax(keywords):
```
Ping https://github.com/in-toto/in-toto/pull/48#discussion_r88950672

SantiagoTorres replied:
Hmm? I don't know how that comment relates to this.

SantiagoTorres replied:
ping @lukpueh

lukpueh replied:
The comment relates to `check_matchrule_syntax`'s docstring.
Name, Description, Returns, Raises != Purpose, Arguments, Exceptions, Side Effects, Return Value

lukpueh commented on test/models/test_inspection.py:
```diff
@@ -58,15 +65,20 @@ def test_wrong_material_matchrules(self):
       self.inspection._validate_product_matchrules()
+
+    with self.assertRaises(FormatError):
       self.assertFalse(self.inspection.validate())
```
Don't need assertFalse here.

lukpueh commented on test/models/test_inspection.py:
```diff
@@ -58,15 +65,20 @@ def test_wrong_material_matchrules(self):
       self.inspection._validate_product_matchrules()
+
+    with self.assertRaises(FormatError):
       self.assertFalse(self.inspection.validate())
```
Don't need assertFalse here

lukpueh commented on test/models/test_inspection.py:
```diff
@@ -26,30 +26,37 @@ class TestInspectionValidator(unittest.TestCase):
       self.inspection._validate_material_matchrules()
+
+    with self.assertRaises(FormatError):
       self.assertFalse(self.inspection.validate())
```
Don't need assertFalse here

lukpueh commented on test/models/test_inspection.py:
```diff
@@ -26,30 +26,37 @@ class TestInspectionValidator(unittest.TestCase):
       self.inspection._validate_material_matchrules()
+
+    with self.assertRaises(FormatError):
       self.assertFalse(self.inspection.validate())
```
Don't need assertFalse here

lukpueh commented on test/models/test_inspection.py:
```diff
@@ -26,30 +26,37 @@ class TestInspectionValidator(unittest.TestCase):
       self.inspection._validate_type()
+
+    with self.assertRaises(FormatError):
       self.assertFalse(self.inspection.validate())
```
Don't need assertFalse here

lukpueh commented on test/models/test_inspection.py:
```diff
@@ -75,13 +87,13 @@ def test_wrong_product_matchrules(self):
     self.assertFalse(self.inspection.validate())
```
I wonder why the test does not fail here. Shouldn't validate always return `None`?

lukpueh commented on test/models/test_step.py:
```diff
@@ -27,44 +27,53 @@ class TestStepValidator(unittest.TestCase):
       self.step._validate_type()
+
+    with self.assertRaises(FormatError):
       self.assertFalse(self.step.validate())
```
No need for assertFalse

lukpueh commented on test/models/test_step.py:
```diff
@@ -27,44 +27,53 @@ class TestStepValidator(unittest.TestCase):
       self.step._validate_material_matchrules()
+
+    with self.assertRaises(FormatError):
       self.assertFalse(self.step.validate())
```
No need for assertFalse

lukpueh commented on test/models/test_step.py:
```diff
@@ -27,44 +27,53 @@ class TestStepValidator(unittest.TestCase):
       self.step._validate_material_matchrules()
+
+    with self.assertRaises(FormatError):
       self.assertFalse(self.step.validate())
```
No need for assertFalse

lukpueh commented on test/models/test_step.py:
```diff
@@ -73,49 +82,54 @@ def test_wrong_material_matchrules(self):
       self.step._validate_product_matchrules()
+
+    with self.assertRaises(FormatError):
       self.assertFalse(self.step.validate())
```
No need for assertFalse

lukpueh commented on test/models/test_step.py:
```diff
@@ -73,49 +82,54 @@ def test_wrong_material_matchrules(self):
       self.step._validate_product_matchrules()
+
+    with self.assertRaises(FormatError):
       self.assertFalse(self.step.validate())
```
No need for assertFalse

lukpueh commented on test/models/test_step.py:
```diff
@@ -73,49 +82,54 @@ def test_wrong_material_matchrules(self):
     self.assertFalse(self.step.validate())
```
I wonder why the test does not fail here. Shouldn't validate always return `None`?

lukpueh commented on test/models/test_step.py:
```diff
@@ -73,49 +82,54 @@ def test_wrong_material_matchrules(self):
     # for more thorough tests, check the test_matchrule.py module
     self.step.product_matchrules = [["CREATE", "foo"]]
     self.step._validate_product_matchrules()
     self.assertFalse(self.step.validate())
```
I wonder why the test does not fail here. Shouldn't validate always return `None`?

lukpueh reviewed (changes requested):
A few very small change requests. Please fix and we are good to go!

SantiagoTorres committed:
FIX:test:models: removes unnecessary assertFalse

Due to a legacy API design, I had a bunch of self.assertFalse that
didn't do anything. Some of them may have caused the tests to pass when
they shouldn't.

SantiagoTorres committed:
STY:toto:matchrule_validators: MATCH variable rename

The MATCH matchrule handler had two variables, target and new_target.
These names didn't match the ones for the fields in the specification or
anywhere else in the code. Rename these variables to match the
specification and the rest of the codebase (target -> path_pattern,
new_target -> target_path_pattern).

SantiagoTorres committed:
DOC:toto:matchrule_validators: s/Raises/Exceptions

The docstrings were in the wrong format, as the field name for the
exceptions thrown by functions is "exceptions" not "raises"

lukpueh commented on toto/matchrule_validators.py:
```diff
@@ -83,7 +83,7 @@ def check_matchrule_syntax(keywords):
     None
 
-  <Raises>
+  <Exceptions>
```
Almost there. :) Please remove `<Name>`, change `<Description>` to `<Purpose>` and add `<Arguments>`.

SantiagoTorres replied:
lol damnit, I totally forgot these...

SantiagoTorres committed:
DOC:toto:matchrule_validators: fix docstrings

The docstring style had the wrong fields (see parent commit for an
example)
